### PR TITLE
Fix #983 by saving future and checking for + raising any exceptions

### LINKED ISSUE #983: Exception never retrieved when using MultiThreadedExecutor and Timers
## Bug report

**Required Info:**

- Operating System:
Ubuntu 20.04.4
- Installation type:
binaries
- Version or commit hash:
foxy
- DDS implementation:
default
- Client library (if applicable):
rclpy

#### Steps to reproduce issue
I am encountering an issue with exceptions being thrown in timer callbacks using `MultiThreadedExecutor` going silent.

The sample shows that when a an exception is generated in a timer callback using the `SingleThreadExecutor`, implicitly I believe, the exception rises if uncaught, as expected.

When using the `MultiThreadedExecutor` the exception gets lost and silently fails, sometimes causing head scratching.
```
#!/usr/bin/env python3
import rclpy
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node


class TestNode(Node):
    def __init__(self):
        super().__init__('zf9p')
        self.tmr = self.create_timer(1, self.tmr_callback)

    def tmr_callback(self):
        1/0
        print('Timer callback fired')


def single_executor():
    test_node = TestNode()

    while rclpy.ok() is True:
        rclpy.spin(test_node)

    test_node.destroy_timer(test_node.tmr)


def multi_executor():
    test_node = TestNode()

    # Moved to multi threaded executor so none of Actions block publishing of msgs
    executor = MultiThreadedExecutor(num_threads=1)
    executor.add_node(test_node)

    while rclpy.ok() is True:
        executor.spin()

    test_node.destroy_timer(test_node.tmr)


if __name__ == '__main__':
    rclpy.init()

    # single_executor()  # un comment to verify proper operation 
    multi_executor()

    rclpy.shutdown()
```

#### Expected behavior
Unhandled exception to rise to the top
```
$ python3 timer_example
Traceback (most recent call last):
  ...
  File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 343, in _execute_timer
    await await_or_execute(tmr.callback)
  File "/opt/ros/foxy/lib/python3.8/site-packages/rclpy/executors.py", line 118, in await_or_execute
    return callback(*args)
  File "timer_example", line 13, in tmr_callback
    1/0
ZeroDivisionError: division by zero
```

#### Actual behavior
Exception seems to silently fail not handled, but after 10 or 30 seconds the following message does show.
```
$ python3 timer_example
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
```

#### Additional information
This looks to be similar to https://github.com/ros2/rclpy/issues/296, but not for certain.

### PR DESCRIPTION
**Pull request by Achllle**
Fixes #983 while preserving functionality as reported in #1016 

Implementation follows @fujitatomoya 's [recommendation](https://github.com/ros2/rclpy/issues/983#issuecomment-1271782711). Rather than save the executor's returned future, I save the handler since the future doesn't seem to have access to the exception.

### CODE REVIEW

**Comment by fujitatomoya** [comment]
@Achllle could you confirm this fixes https://github.com/fujitatomoya/ros2_test_prover/blob/master/prover_rclpy/src/rclpy_983.py ? as far as i see, it still does not catch the exception.

btw, regression test is already in upstream, https://github.com/ros2/rclpy/pull/1044. this can confirm that there is no regression such as https://github.com/ros2/rclpy/issues/1039.

**Comment by Achllle** [comment]
Hmm, it seems like it's only a partial fix. 
Without this fix, the test case will 'hang' for ~10s, then burst out a stream of
```
The following exception was never retrieved: division by zero
The following exception was never retrieved: division by zero
...
```
and then repeat 10 seconds later, each time without any traceback.

With this fix, the script will fail immediately upon raising the `ZeroDivisionError` with traceback:
```
Traceback (most recent call last):
  File "/workspaces/dockerfiles/rclpy_src/src/rclpy_983.py", line 48, in <module>
    main()
  File "/workspaces/dockerfiles/rclpy_src/src/rclpy_983.py", line 42, in main
    multi_executor()
  File "/workspaces/dockerfiles/rclpy_src/src/rclpy_983.py", line 33, in multi_executor
    executor.spin()
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 280, in spin
    self.spin_once()
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 789, in spin_once
    self._spin_once_impl(timeout_sec)
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 785, in _spin_once_impl
    raise future._exception
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/task.py", line 239, in __call__
    self._handler.send(None)
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 424, in handler
    await call_coroutine(entity, *arg)
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 334, in _execute_timer
    await await_or_execute(tmr.callback)
  File "/workspaces/rolling/install/rclpy/local/lib/python3.10/dist-packages/rclpy/executors.py", line 108, in await_or_execute
    return callback(*args)
  File "/workspaces/dockerfiles/rclpy_src/src/rclpy_983.py", line 12, in tmr_callback
    1/0
ZeroDivisionError: division by zero
The following exception was never retrieved: division by zero
```
but will still produce a `The following exception was never retrieved: division by zero`. I will dig in a bit more to see if I can rectify that, in the meantime any suggestions are welcome.

**Comment by Achllle** [comment]
@fujitatomoya Can you take a look again? The latest commit should fix the `The following exception was never retrieved: division by zero` issue by calling `future.result()`. Tested on #983 test.

File: rclpy/rclpy/executors.py
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:758]: this meant to be internal use only?

```suggestion
        self._futures = []
```

**Review by fujitatomoya — COMMENTED** [review]
one minor comment needs to be addressed, overall lgtm.

btw, i confirmed that this can fix the issue to raise the exception right away.

**Comment by fujitatomoya** [comment]
@wjwwood minor fix, this can address the issue https://github.com/ros2/rclpy/issues/983, can you take a look?

File: rclpy/rclpy/executors.py
Comment (Achllle) [inline rclpy/rclpy/executors.py:779]:
```suggestion
            self._futures.append(handler)
```

File: rclpy/rclpy/executors.py
Comment (Achllle) [inline rclpy/rclpy/executors.py:780]:
```suggestion
            for future in self._futures:  # check for any exceptions
```

File: rclpy/rclpy/executors.py
Comment (Achllle) [inline rclpy/rclpy/executors.py:782]:
```suggestion
                    self._futures.remove(future)
```

**Review by fujitatomoya — APPROVED** [review]
lgtm

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18046/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/12575/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/18691/)

**Comment by tonynajjar** [comment]
I happened to stumble on the same bug, I tested the fix (on top of humble) and it worked for me. Looking forward for the merge and backport please :grin:

**Comment by Achllle** [comment]
@fujitatomoya @wjwwood Anything I can do to move this along? Linux-aarch64 CI failure looks unrelated to me.

**Comment by adityapande-1995** [comment]
@Achllle I'll trigger CI once more, and if that passes, I'll merge this.

**Comment by clalancette** [comment]
For what it is worth, we shouldn't merge things in with failing CI.  In this particular case it didn't matter (CI was happy overnight), but it isn't our usual development workflow.

**Comment by fujitatomoya** [comment]
@Mergifyio backport humble

**Comment by fujitatomoya** [comment]
@Mergifyio backport foxy

**Comment by jdlangs** [comment]
This works but it relies on another callback being available for execution before the exception is raised since `wait_for_ready_callbacks` can block indefinitely. If you have a node that just provides services, for example, you won't get the exception until someone makes another service call after the failing one.

Would it make sense to modify this to use a timeout and keep checking the futures on a regular basis?

**Comment by froly** [comment]
@Achllle @fujitatomoya I believe there's a typo  in this patch
```
self._executor.submit(handler)
self._futures.append(handler)
```
should be
```
future = self._executor.submit(handler)
self._futures.append(future)
```

we want to store and check the `Future` returned by `executor.submit()` to make this work and propagate an exception

**Pull request merged**
